internal/app/machined/pkg/controllers/k8s/kubelet_service.go

```diff
@@ -297,3 +324,86 @@ func updateKubeconfig(newEndpoint *url.URL) error {
 
 	return clientcmd.WriteToFile(*config, constants.KubeletKubeconfig)
 }
+
+// refreshKubeletCerts checks if the existing kubelet certificates match the node hostname.
+// If they don't match, it clears the certificate directory and the removes kubelet's kubeconfig so that
+// they can be regenerated next time kubelet is started.
+func (ctrl *KubeletServiceController) refreshKubeletCerts(hostname string) error {
+	cert, err := ctrl.readKubeletCertificate()
+	if err != nil {
+		return err
+	}
+
+	if cert == nil {
+		return nil
+	}
+
+	valid := slices.Contains(cert.DNSNames, func(name string) bool {
+		return name == hostname
+	})
+
+	if valid {
+		// certificate looks good, no need to refresh
+		return nil
+	}
+
+	dir, err := os.ReadDir(constants.KubeletPKIDir)
+	if errors.Is(err, os.ErrNotExist) {
+		return nil
+	}
+
+	if err != nil {
+		return err
+	}
+
+	// clear the pki directory
+	for _, entry := range dir {
```

Comment:
do we need to go over each entry or is it okay to remove the whole `constants.KubeletPKIDir`?

Reply:
hm, it seemed to be necessary but that was before I added `os.Remove(constants.KubeletKubeconfig)` afterwards. let me give it another try now.

Reply:
yep, it works fine when deleted. updated the code.